src/sapling/sapling_operation.cpp

```diff
@@ -143,12 +161,20 @@ OperationResult SaplingOperation::build()
         }
 
         // Add outputs
+        bool fFirst = false;
         for (const SendManyRecipient &t : recipients) {
             if (t.IsTransparent()) {
-                txBuilder.AddTransparentOutput(CTxOut(t.transparentRecipient->nAmount, t.transparentRecipient->scriptPubKey));
+                CAmount amount = t.transparentRecipient->nAmount;
+                if (t.transparentRecipient->fSubtractFeeFromAmount) {
+                    subtractFee(fFirst, amount, nFeeRet, nSubtractFeeFromAmount);
+                }
+                txBuilder.AddTransparentOutput(CTxOut(amount, t.transparentRecipient->scriptPubKey));
             } else {
                 const auto& address = t.shieldedRecipient->address;
-                const CAmount& amount = t.shieldedRecipient->amount;
+                CAmount amount = t.shieldedRecipient->amount;
+                if (t.shieldedRecipient->fSubtractFeeFromAmount) {
+                    subtractFee(fFirst, amount, nFeeRet, nSubtractFeeFromAmount);
+                }
```

Comment:
The first receiver will never pay the remainder, `fFirst` is always false.
Need to initialize it to true (line 164), so the `subtractFee` function internally sets it to false after the first loop cycle.

We should try to avoid the update of external variables inside utility functions. It ends up being an error-prone design.

I have worked a bit on the problem, abstracting `CRecipient/ShieldedRecipient` so we can clean some code duplication and don't be forced to do something like the `subtractFee` static function. Check it here https://github.com/furszy/PIVX/commits/202104_subtractfeefromamount.

Reply:
Good catch, fixed initialization of `fFirst`.

Your abstraction looks great, but it seems there's some issue with uninitialized `Optional` in some unit-test, consistently on GA (https://github.com/furszy/PIVX/runs/2742030021). I couldn't reproduce it locally though.
In any case, I think it's a good refactoring, that deserves its own PR.

Reply:
I forgot to answer this comment.
The GA job that you are watching is from an old branch, using optional instead of unique pointer. I pushed it half an hour after my comment here, probably 3 hours before yours. Probably GitHub played with your hearth or you had open the job of the previous branch and didn't see the update. The current branch there is passing nicely.

Side from that, no problem in pushing it in another PR, i did it mainly because i'm not happy with the `subtractFee` static function. The rest looks good.